lib/bolt/fiber_executor.rb

```diff
@@ -0,0 +1,168 @@
+      @id = 0
+      @plan_futures = []
+    end
+
+    # Whether there is more than one fiber running in parallel.
+    #
+    def in_parallel
+      plan_futures.length > 1
+    end
+
+    # Creates a new Puppet scope from the current Plan scope so that variables
+    # can be used inside the block and won't interact with the outer scope.
+    # Then creates a new Fiber to execute the block, wraps the Fiber in a
+    # Bolt::PlanFuture, and returns the Bolt::PlanFuture. Because the main plan
+    # doesn't need the scoping and is called by the plan compiler in PAL, this
+    # is separate from `add_future` which adds the future to the list of
+    # futures to round_robin on. This is so that the main plan can create it's
+    # own PlanFuture and add it, while all other PlanFutures are created here.
+    #
+    def create_future(scope, block, name: nil, object: nil)
+      # Save existing variables to the new scope before starting the future
+      # itself so that if the plan returns before the backgrounded block
+      # starts, we still have the variables.
+      newscope = Puppet::Parser::Scope.new(scope.compiler)
+      local = Puppet::Parser::Scope::LocalScope.new
+
+      # Compress the current scopes into a single vars hash to add to the new scope
+      scope.to_hash.each_pair { |k, v| local[k] = v }
+      newscope.push_ephemerals([local])
+
+      # Create a new Fiber that will execute the provided block.
+      future = Fiber.new do
+        # Catch 'return' calls inside the block
+        result = catch(:return) do
+          # If called with `parallelize()`, provide the argument from the block.
+          args = object.nil? ? {} : { block.parameters[0][1].to_s => object }
+          # Execute the block. Individual plan steps in the block will yield
+          # the Fiber if they haven't finished, so all this needs to do is run
+          # the block.
+          block.closure.call_by_name_with_scope(newscope, args, true)
+        end
+
+        # If we got a return from the block, get it's value
+        # Otherwise the result is the last line from the block
+        result = result.value if result.is_a?(Puppet::Pops::Evaluator::Return)
+
+        # Validate the result is a PlanResult
+        unless Puppet::Pops::Types::TypeParser.singleton.parse('Boltlib::PlanResult').instance?(result)
+          raise Bolt::InvalidParallelResult.new(result.to_s, *Puppet::Pops::PuppetStack.top_of_stack)
+        end
+
+        result
+      rescue Puppet::PreformattedError => e
+        if e.cause.is_a?(Bolt::Error)
+          e.cause
+        else
+          raise e
+        end
+      end
+
+      # PlanFutures are assigned an ID, which is just a global incrementing
+      # integer. The main plan should always have ID 0.
+      @id += 1
+      Bolt::PlanFuture.new(future, @id, name)
+    end
+
+    # Register a PlanFuture with the FiberExecutor to be executed while round_robining.
+    def add_future(future)
```

Comment:
separating `create` and `add` in the API doesn't make much sense. Is it only there for top level that doesn't need to copy the scope? Is there a clean way to make a single create/add method to expose as API?

Reply:
It's also because the plan and the block are called pretty different - the main plan is called through the compiler directly, whereas the code block is a `Puppet::Pops::Evaluator::PuppetProc` that has to be called with the new scope. I think I can have create_future accept a block that goes in the fiber perhaps...

Reply:
Ok, I think what I've got not makes a lot more sense and provides a cleaner API